src/transactions/ClaimClaimableBalanceOpFrame.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "transactions/ClaimClaimableBalanceOpFrame.h"
+#include "crypto/SHA.h"
+#include "ledger/LedgerTxn.h"
+#include "ledger/LedgerTxnEntry.h"
+#include "ledger/LedgerTxnHeader.h"
+#include "ledger/TrustLineWrapper.h"
+#include "transactions/TransactionUtils.h"
+
+namespace stellar
+{
+
+ClaimClaimableBalanceOpFrame::ClaimClaimableBalanceOpFrame(
+    Operation const& op, OperationResult& res, TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mClaimClaimableBalance(mOperation.body.claimClaimableBalanceOp())
+{
+}
+
+ThresholdLevel
+ClaimClaimableBalanceOpFrame::getThresholdLevel() const
+{
+    return ThresholdLevel::LOW;
+}
+
+bool
+ClaimClaimableBalanceOpFrame::isVersionSupported(uint32_t protocolVersion) const
+{
+    return protocolVersion >= 14;
+}
+
+bool
+validatePredicate(ClaimPredicate const& pred, TimePoint closeTime)
+{
+    switch (pred.type())
+    {
+    case CLAIM_PREDICATE_AND:
+    {
+        auto const& andPredicates = pred.andPredicates();
+        assert(andPredicates.size() == 2);
+
+        return validatePredicate(andPredicates[0], closeTime) &&
+               validatePredicate(andPredicates[1], closeTime);
+    }
+
+    case CLAIM_PREDICATE_OR:
+    {
+        auto const& orPredicates = pred.orPredicates();
+        assert(orPredicates.size() == 2);
+
+        return validatePredicate(orPredicates[0], closeTime) ||
+               validatePredicate(orPredicates[1], closeTime);
+    }
+
+    case CLAIM_PREDICATE_BEFORE_ABSOLUTE_TIME:
+        return pred.absBefore() > closeTime;
+    case CLAIM_PREDICATE_AFTER_ABSOLUTE_TIME:
+        return pred.absAfter() < closeTime;
+    case CLAIM_PREDICATE_BEFORE_RELATIVE_TIME:
+    case CLAIM_PREDICATE_AFTER_RELATIVE_TIME:
+        // TODO:shouldn't be able to get here... is aborting fine?
```

Comment:
Added a throw here for now. Using a different `ClaimPredicate` sounds interesting though. I'm fine with either solution.

Reply:
btw in other places in the code we use `abort()` in this situation - I agree thought that in general it's better practice to throw instead instead of crashing (maybe worth creating an issue for this to clean up)

-> default case should throw